x-pack/plugins/reporting/public/panel_actions/get_csv_panel_action.tsx

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import dateMath from '@elastic/datemath';
+import { i18n } from '@kbn/i18n';
+
+import { ContextMenuAction, ContextMenuActionsRegistryProvider } from 'ui/embeddable';
+import { PanelActionAPI } from 'ui/embeddable/context_menu_actions/types';
+import { kfetch } from 'ui/kfetch';
+import { toastNotifications } from 'ui/notify';
+
+const API_BASE_URL = '/api/reporting/v1/generate/immediate/csv/saved-object/';
+
+class GetCsvReportPanelAction extends ContextMenuAction {
+  constructor() {
+    super(
+      {
+        displayName: i18n.translate('xpack.reporting.dashboard.downloadCsvPanelTitle', {
+          defaultMessage: 'Download CSV',
+        }),
+        id: 'downloadCsvReport',
+        parentPanelId: 'mainMenu',
+      },
+      {
+        icon: 'document',
+      }
+    );
+  }
+
+  public async generateJobParams({ searchEmbeddable }: { searchEmbeddable: any }) {
+    const adapters = searchEmbeddable.getInspectorAdapters();
+    if (!adapters) {
+      return {};
+    }
+
+    if (adapters.requests.requests.length === 0) {
+      return {};
+    }
+
+    return searchEmbeddable.searchScope.searchSource.getSearchRequestBody();
+  }
+
+  public isVisible = (panelActionAPI: PanelActionAPI): boolean => {
+    const { embeddable } = panelActionAPI;
+
```

Comment:
Had to go with duck-type-typing here since it was proving to be impossible to import the SavedSearchEmbedable component :/

Reply:
Ok, so the actual class we want is from `src/legacy/core_plugins/kibana/public/discover/embeddable/search_embeddable.ts`, which is an extension of the Embeddable class. That's why we have to check for the `savedSearch` property to make sure it's valid?

In that case, I think this could use a comment to marker where we might be able to improve things down the road.